File: python/leetcode/leetcode3286.py
# https://leetcode.com/problems/find-a-safe-walk-through-a-grid

import heapq

class Solution:
    def findSafeWalk(self, grid: List[List[int]], health: int) -> bool:
        m, n = len(grid), len(grid[0])
        def dijkstra(start: Tuple[int, int]) -> int:
            heap = []
            distance = [[float('inf')] * n for _ in range(m)]
            heapq.heappush(heap, (grid[start[0]][start[1]], start))
            distance[start[0]][start[1]] = grid[start[0]][start[1]]
            
            moves = [(-1, 0), (0, 1), (1, 0), (0, -1)]
            while heap:
                dist, (row, col) = heapq.heappop(heap)
                if dist > distance[row][col]:
                    continue
                for dr, dc in moves:
                    newRow, newCol = row + dr, col + dc
                    if newRow < 0 or newRow >= m or newCol < 0 or newCol >= n:
                        continue
                    if dist + grid[newRow][newCol] < distance[newRow][newCol]:
                        distance[newRow][newCol] = dist + grid[newRow][newCol]
                        heapq.heappush(heap, (dist + grid[newRow][newCol], (newRow, newCol)))
            return distance[m - 1][n - 1]
        return dijkstra(start = (0, 0)) < health

File: swift/leetcode/leetcode3286.swift
// https://leetcode.com/problems/find-a-safe-walk-through-a-grid

class Solution {
    typealias Location = (row: Int, col: Int)
    func findSafeWalk(_ grid: [[Int]], _ health: Int) -> Bool {
        let (m, n) = (grid.count, grid.first!.count)
        func bfs(start: Location) -> Int {
            var queue = [Location]()
            var visited = [[Int]](repeating: [Int](repeating: -1, count: n), count: m)
            queue.append(start)
            visited[start.row][start.col] = grid[start.row][start.col]
            while !queue.isEmpty {
                let (row, col) = queue.removeFirst()
                for (dr, dc) in [(-1, 0), (0, 1), (1, 0), (0, -1)] {
                    let (newRow, newCol) = (row + dr, col + dc)
                    guard 0..<m ~= newRow && 0..<n ~= newCol else { continue }
                    guard visited[newRow][newCol] == -1 else { continue }
                    if grid[newRow][newCol] == 0 {
                        visited[newRow][newCol] = visited[row][col]
                        queue.insert((newRow, newCol), at: 0)
                    } else if grid[newRow][newCol] == 1 {
                        visited[newRow][newCol] = visited[row][col] + 1
                        queue.append((newRow, newCol))
                    }
                }
            }
            return visited[m - 1][n - 1]
        }
        return health > bfs(start: (0, 0))
    }
}
